### PR TITLE
Feat/multichain payment

### PR DESCRIPTION
**Pull request by volodymyr-basiuk**

### CODE REVIEW

**Review by daveroga — APPROVED** [review]

File: contracts/payment/MCPayment.sol [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    error InvalidSignature(string message);
+    error PaymentError(string message);
+    error WithdrawError(string message);
+
+    /**
+     * @dev Initialize the contract
+     */
+    function initialize(address owner, uint8 ownerPercentage) public initializer {
+        MCPaymentStorage storage $ = _getMCPaymentStorage();
+        $.ownerPercentage = ownerPercentage;
```
Comment (AndriianChestnykh) [inline contracts/payment/MCPayment.sol:61]: Worth doing any sanity checks for ownerPercentage? e.g in between 0 and 100

File: contracts/payment/MCPayment.sol [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+        if (err != ECDSA.RecoverError.NoError || recovered != paymentData.recipient) {
+            revert InvalidSignature("MCPayment: invalid signature");
+        }
+    }
+
+    function getMyBalance() public view returns (uint256) {
+        MCPaymentStorage storage $ = _getMCPaymentStorage();
+        return $.balance[_msgSender()];
+    }
```
Comment (AndriianChestnykh) [inline contracts/payment/MCPayment.sol:128]: Does it really work if you call the method without transaction? The will be no signature so no msg.sender. Maybe `getBalance(address)` is better? Besides, you can ask for any address that way.

File: helpers/DeployHelper.ts [hunk 3]
```diff
@@ -1085,6 +1086,59 @@ export class DeployHelper {
...
+    const MCPaymentFactory = await ethers.getContractFactory("MCPayment");
+    const Create2AddressAnchorFactory = await ethers.getContractFactory("Create2AddressAnchor");
+
+    let mcPayment;
+    if (deployStrategy === "create2") {
+      this.log("deploying with CREATE2 strategy...");
+
+      // Deploying MCPayment contract to predictable address but with dummy implementation
+      mcPayment = (
+        await ignition.deploy(MCPaymentModule, {
```
Comment (AndriianChestnykh) [inline helpers/DeployHelper.ts:1108]: `MCPaymentProxyModule` but not `MCPaymentModule` should be here. You don't need `MCPaymentModule` at all.

File: ignition/modules/mcPayment.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import { buildModule } from "@nomicfoundation/hardhat-ignition/modules";
+import { contractsInfo } from "../../helpers/constants";
+
+/**
+ * This is the first module that will be run. It deploys the proxy and the
+ * proxy admin, and returns them so that they can be used by other modules.
+ */
```
Comment (AndriianChestnykh) [inline ignition/modules/mcPayment.ts:7]: Do we need this comment?

File: ignition/modules/mcPayment.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  // Here we're using m.contractAt(...) a bit differently than we did above.
+  // While we're still using it to create a contract instance, we're now telling Hardhat Ignition
+  // to treat the contract at the proxy address as an instance of the Demo contract.
+  // This allows us to interact with the underlying Demo contract via the proxy from within tests and scripts.
+  const mcPayment = m.contractAt("MCPayment", proxy);
+
+  // Return the contract instance, along with the original proxy and proxyAdmin contracts
+  // so that they can be used by other modules, or in tests and scripts.
+  return { mcPayment, proxy, proxyAdmin };
+});
```
Comment (AndriianChestnykh) [inline ignition/modules/mcPayment.ts:47]: We don't need this module

File: test/payment/mc-payment.test.ts [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    // issuer 1 balance should be 0
+    const issuer1BalanceAfterWithdrow = await payment.connect(issuer1Signer).getMyBalance();
+    expect(issuer1BalanceAfterWithdrow).to.be.eq(0);
+
+    const receipt = await issuerWithdrawTx.wait();
+    const gasSpent = receipt!.gasUsed * receipt!.gasPrice;
+
+    expect(await ethers.provider.getBalance(issuer1Signer.address)).to.be.eq(
+      issuerBalanceBeforeWithdraw + issuer1BalanceInContract - gasSpent,
+    );
```
Comment (AndriianChestnykh) [inline test/payment/mc-payment.test.ts:97]: There are convenient helpers, which take into account both the gas spent and the balance change.

File: test/payment/mc-payment.test.ts [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    const isPaymentDone = await payment.isPaymentDone(issuer1Signer.address, 25);
+    expect(isPaymentDone).to.be.true;
+
+    // issuer withdraw
+    const issuerBalanceBeforeWithdraw = await ethers.provider.getBalance(issuer1Signer.address);
+    const issuer1BalanceInContract = await payment.connect(issuer1Signer).getMyBalance();
+    expect(issuer1BalanceInContract).to.be.eq(90);
+    const issuerWithdrawTx = await payment.connect(issuer1Signer).issuerWithdraw();
+    // issuer 1 balance should be 0
+    const issuer1BalanceAfterWithdrow = await payment.connect(issuer1Signer).getMyBalance();
```
Comment (AndriianChestnykh) [inline test/payment/mc-payment.test.ts:89]: As I explained above, this can be 0 all the time (even if not) if asked as a call but not transaction

**Review by AndriianChestnykh — APPROVED** [review]

**Review by daveroga — APPROVED** [review]

**Review by vmidyllic — APPROVED** [review]

**Pull request merged**